File: src/firefly/java/edu/caltech/ipac/firefly/util/BrowserInfo.java
/*
 * License information at https://github.com/Caltech-IPAC/firefly/blob/master/License.txt
 */
package edu.caltech.ipac.firefly.util;

import edu.caltech.ipac.util.StringUtils;

/**
 * Date: Feb 22, 2009
 *
 /**
 *
 * Get info about the Browser. Should be used on the server side or when
 * wrapped by Browser Info
 *
 * @author Trey
 * @version $Id: BrowserInfo.java,v 1.17 2012/09/26 22:08:20 roby Exp $
 */
public class BrowserInfo {


    public static final int UNKNOWN_VERSION= -1;
    private static final String SEAMONKEY_KEY= "seamonkey/";
    private static final String FIREFOX_KEY= "firefox/";
    private static final String SIMPLE_VERSION_KEY = "version/";
    private static final String CHROME_KEY= "chrome/";

    private Browser _browser;
    private Platform _platform;
    private int _majorVersion;
    private int _minorVersion= UNKNOWN_VERSION;
    private final String _userAgent;
    private boolean _allRecognized;

    public BrowserInfo(String userAgent) {
        _userAgent= userAgent;
        _platform= getPlatform(userAgent);
        _allRecognized = (_platform!=Platform.UNKNOWN);
        evaluateBrowser(_platform);
    }

    public Browser getBrowserType() { return _browser; }
    public boolean isIE() { return isBrowser(Browser.IE); }
    public boolean isSafari() { return isBrowser(Browser.SAFARI); }
    public boolean isFirefox() { return isBrowser(Browser.FIREFOX); }
    public boolean isChrome() { return isBrowser(Browser.CHROME); }

    public boolean isBrowser(Browser browser) { return isBrowser(browser, UNKNOWN_VERSION); }
    public boolean isBrowser(Browser browser, int version) {
        boolean retval= false;
        if (browser==_browser) {
            retval= (version==UNKNOWN_VERSION || version==_majorVersion);
        }
        return retval;
    }


    public boolean isVersionAtLeast(Browser browser, int version, int minor) {
        boolean retval= false;
        if (browser==_browser) {
            if (version!=UNKNOWN_VERSION) {
                if (_majorVersion>version) {
                    retval= true;
                }
                if (_majorVersion==version) {
                    if (_minorVersion!=UNKNOWN_VERSION) {
                        if (_minorVersion>=minor)  {
                            retval= true;
                        }
                    }
                }
            }
        }
        return retval;
    }


    public boolean isVersionAtLeast(Browser browser, int version) {
        boolean retval= false;
        if (browser==_browser) {
            if (version!=UNKNOWN_VERSION) {
                if (_majorVersion>=version) {
                    retval= true;
                }
            }
        }
        return retval;
    }

    public boolean isVersionBefore(Browser browser, int version, int minor) {
        return !isVersionAtLeast(browser,version,minor);
    }

    public boolean isVersionBefore(Browser browser, int version) {
        return !isVersionAtLeast(browser, version);
    }


    public boolean getSupportsCSS3() {
        boolean retval=  isBrowser(Browser.CHROME);

        if (!retval) {
            retval= isBrowser(Browser.SAFARI) && _majorVersion>=5;
        }

        if (!retval) {
            retval= isTouchInput();
        }

        if (!retval) {
            retval= isBrowser(Browser.FIREFOX) && _majorVersion>3;
        }

        if (!retval) {
            retval= isBrowser(Browser.OPERA) && _majorVersion>=10;
        }

        if (!retval) {
            retval= isIE() && _majorVersion>=9;
        }

        return retval;

    }

    public boolean getSupportsShadows() {
        boolean retval=  true;

        if (isIE() && isVersionBefore(Browser.IE,9)) {
            retval= false;
        }
        else if (isSafari() && isVersionBefore(Browser.SAFARI,5,1)) {
            retval= false;
        }

        return retval;

    }

    public boolean getSupportsCORS() {
        boolean retval=  true;
        if (isFirefox() && isVersionBefore(Browser.FIREFOX,3,5)) {
            retval= false;
        }
        else if (isSafari() && isVersionBefore(Browser.SAFARI,4)) {
            retval= false;
        }
        else if (isIE() && isVersionBefore(Browser.IE,10)) {
            retval= false;
        }
        else if (isBrowser(Browser.OPERA) && isVersionBefore(Browser.OPERA,12)) {
            retval= false;
        }
        return retval;

    }


    public boolean isPlatform(Platform platform) { return  (platform == _platform); }
    public int getMajorVersion()  { return _majorVersion; }
    public int getMinorVersion()  { return _minorVersion; }
    public String getPlatformDesc() { return _platform.getDesc(); }

    public boolean isAllRecognized() { return _allRecognized; }

    public String getBrowserString() { return _browser.getDesc(); }

    public String getVersionString() {
        String retval;
        if (_minorVersion!=UNKNOWN_VERSION) retval= _majorVersion + "." +_minorVersion;
        else                                retval= _majorVersion+"";
        return retval;
    }

    public String getBrowserDesc() {
        String retval;
        if (_minorVersion!=UNKNOWN_VERSION) {
            retval= StringUtils.pad(12, getBrowserString())+
                    "Version: " + StringUtils.pad(15, _majorVersion + "." +_minorVersion);
        }
        else {
            retval= StringUtils.pad(12, getBrowserString())+ "Version: " + StringUtils.pad(15,_majorVersion+"");
        }
        return retval;
    }

    public boolean isTouchInput() {
        return _platform==Platform.IPAD     ||
               _platform==Platform.IPHONE   ||
               _platform==Platform.ANDROID;
    }

    public boolean canSupportAnimation() {
        boolean retval= false;
        switch (_browser) {
            case FIREFOX:
                retval= (_majorVersion>=4) || (_majorVersion==3 && _minorVersion>=5);
                break;
            case SAFARI:
                retval= (_majorVersion>=5);
                break;
            case IE:
                retval= (_majorVersion>=10);
                break;
            case OPERA:
                retval= (_majorVersion>=12);
                break;
            case CHROME:
                retval= true;
                break;
        }
        return retval;

    }

    private void evaluateBrowser(Platform p) {
        _browser= Browser.UNKNOWN;
        _majorVersion= UNKNOWN_VERSION;
        if (_userAgent.contains("msie")) {
            _browser= Browser.IE;
            if (_userAgent.contains("msie 10")) {
                _majorVersion= 10;
            }
            else if (_userAgent.contains("msie 9")) {
                _majorVersion= 9;
            }
            else if (_userAgent.contains("msie 8")) {
                _majorVersion= 8;
            }
            else if (_userAgent.contains("msie 7")) {
                _majorVersion= 7;
            }
            else if (_userAgent.contains("msie 6")) {
                _majorVersion= 6;
            }
        }
        else if (_userAgent.contains(FIREFOX_KEY)) {
            _browser= Browser.FIREFOX;
            _allRecognized= parseVersion(FIREFOX_KEY);
        }
        else if (_userAgent.contains(SEAMONKEY_KEY)) {
            _browser= Browser.SEAMONKEY;
            _allRecognized= parseVersion(SEAMONKEY_KEY);
        }
        else if  (_userAgent.contains("applewebkit")) {
            if (_userAgent.contains(CHROME_KEY)) {
                _browser= Browser.CHROME;
                _allRecognized= parseVersion(CHROME_KEY);
            }
            else if (p==Platform.BLACKBERRY || p==Platform.SYMBIAN_OS || p==Platform.UNKNOWN) {
                _browser= Browser.WEBKIT_GENERIC;
                _allRecognized= parseVersion(SIMPLE_VERSION_KEY);
            }
            else {
                _browser= Browser.SAFARI;
                _allRecognized= parseVersion(SIMPLE_VERSION_KEY);
            }
        }
        else if  (_userAgent.contains("opera")) {
            _browser= Browser.OPERA;
            if (_userAgent.contains("opera/9")) {
                _majorVersion= 9;
                parseVersion(SIMPLE_VERSION_KEY);
            }
            else if (_userAgent.contains("opera/8")) {
                _majorVersion= 8;
            }
            else if (_userAgent.contains("opera/7")) {
                _majorVersion= 7;
            }
            else if (_userAgent.contains("opera/6")) {
                _majorVersion= 6;
            }
        }
        else {
            _browser= Browser.UNKNOWN;
            _allRecognized = false;
        }
    }

    private boolean parseVersion(String key) throws NumberFormatException {
        boolean success= true;
        try {
            int idx= _userAgent.indexOf(key);
            if (idx!=-1) idx+= key.length();
            String ver= _userAgent.substring(idx);

            int len= ver.length();
            boolean found= false;
            int i;
            char c;
            for(i=0; (i<len); i++) {
                c= ver.charAt(i);
                if (!Character.isDigit(c) && c!='.') {
                    found= true;
                    break;
                }
            }
            if (found) ver= ver.substring(0,i);


            String sAry[]= ver.split("\\.");
            _majorVersion= Integer.parseInt(sAry[0]);
            if (StringUtils.allDigits(sAry[1])) {
                _minorVersion= Integer.parseInt(sAry[1]);
            }
        } catch (Exception e) {
            success= false;
        }
        return success;
    }

    private static Platform getPlatform(String ua) {
        Platform platform;
        if  (ua.contains("window")) {
            platform = Platform.WINDOWS;
        }
        else if  (ua.contains("macintosh")) {
            platform = Platform.MAC;
        }
        else if  (ua.contains("ipad")) {
            platform = Platform.IPAD;
        }
        else if  (ua.contains("iphone")) {
            platform = Platform.IPHONE;
        }
        else if  (ua.contains("linux")) {
            if (ua.contains("android")) {
                platform = Platform.ANDROID;
            }
            else {
                platform = Platform.LINUX;
            }
        }
        else if  (ua.contains("solaris")) {
            platform = Platform.SOLARIS;
        }
        else if  (ua.contains("sunos")) {
            platform = Platform.SUNOS;
        }
        else if  (ua.contains("aix")) {
            platform = Platform.AIX;
        }
        else if  (ua.contains("hpux")) {
            platform = Platform.HPUX;
        }
        else if  (ua.contains("freebsd")) {
            platform = Platform.FREE_BSD;
        }
        else if  (ua.contains("symbianos")) {
            platform = Platform.SYMBIAN_OS;
        }
        else if  (ua.contains("j2me")) {
            platform = Platform.J2ME;
        }
        else if  (ua.contains("blackberry")) {
            platform = Platform.BLACKBERRY;
        }
        else {
            platform = Platform.UNKNOWN;
        }
        return platform;
    }

}

File: src/firefly/js/util/BrowserInfo.js
/*
 * License information at https://github.com/Caltech-IPAC/firefly/blob/master/License.txt
 */
import Enum from 'enum';
import {getProp} from 'firefly/util/WebUtil.js';

/**
 * @typedef {Object} Browser
 * enum can be one of
 * @prop FIREFOX
 * @prop SEAMONKEY
 * @prop SAFARI
 * @prop WEBKIT_GENERIC
 * @prop IE
 * @prop OPERA
 * @prop CHROME
 * @prop EDGE
 * @prop UNKNOWN
 * @type {Enum}
 */


/** @type Browser */
export const Browser = new Enum ([
    'FIREFOX', 'SEAMONKEY', 'SAFARI', 'WEBKIT_GENERIC', 'IE', 'OPERA', 'CHROME', 'EDGE', 'UNKNOWN'
]);

/**
 * @typedef {Object} Platform
 * enum can be one of
 * @prop MACOS
 * @prop WINDOWS
 * @prop LINUX
 * @prop SOLARIS
 * @prop SUNOS
 * @prop HPUX
 * @prop AIX
 * @prop IPHONE
 * @prop IPAD
 * @prop ANDROID
 * @prop FREE_BSD
 * @prop SYMBIAN_OS
 * @prop J2ME
 * @prop BLACKBERRY
 * @prop UNKNOWN
 * @type {Enum}
 */

/** @type Platform */
export const Platform = new Enum([
    'MACOS', 'WINDOWS', 'LINUX', 'SOLARIS', 'SUNOS', 'HPUX', 'AIX', 'IPHONE', 'IPAD', 'ANDROID', 'FREE_BSD',
    'SYMBIAN_OS', 'J2ME', 'BLACKBERRY', 'UNKNOWN'
]);

const UNKNOWN_VER= -1;

const BrowserInfo= {
    getBrowserType: () => BrowserInfo.browser,
    isSafari: () => isBrowser(Browser.SAFARI),
    isFirefox: () => isBrowser(Browser.FIREFOX),
    isChrome: () => isBrowser(Browser.CHROME),
    isEdge: () => isBrowser(Browser.EDGE),
    isChromeLike: () => isBrowser(Browser.CHROME) || isBrowser(Browser.EDGE) || isBrowser(Browser.OPERA),
    isBrowser,
    isPlatform: (inPlatform) => inPlatform===BrowserInfo.platform,
    isVersionAtLeast,
    isVersionBefore: (version, minor=UNKNOWN_VER) => !isVersionAtLeast(version,minor),
    getPlatformDesc: () => BrowserInfo.platform.key,
    getBrowserString: () => BrowserInfo.browser.key,
    isTouchInput: () => ([Platform.IPAD,Platform.IPHONE,Platform.ANDROID].includes(BrowserInfo.platform)),
    getBrowserDesc: () => BrowserInfo.browserDesc,
    supportsCssColorMix,
    getMinSupportedVersion,
    isBrowserVersionSupported,
    supportsWebGpu,
    getVersionString: () =>
        BrowserInfo.minorVersion!==UNKNOWN_VER ? `${BrowserInfo.majorVersion}.${BrowserInfo.minorVersion}` : BrowserInfo.majorVersion+'',
    minorVersion: undefined,
    majorVersion: undefined,
    browser: undefined,
};

export default BrowserInfo;


(() => {
    const ua= globalThis?.navigator?.userAgent?.toLowerCase();
    BrowserInfo.userAgent= ua;
    BrowserInfo.platform= evaluatePlatform(ua);
    const {browser,majorVersion,minorVersion} = evaluateBrowser(ua,BrowserInfo.platform);
    BrowserInfo.majorVersion=majorVersion;
    BrowserInfo.minorVersion=minorVersion;
    BrowserInfo.browser=browser;
    BrowserInfo.browserDesc= BrowserInfo.minorVersion!==UNKNOWN_VER ?
            BrowserInfo.getBrowserString()+ ' Version: ' + BrowserInfo.majorVersion + '.' +BrowserInfo.minorVersion :
            BrowserInfo.getBrowserString()+ ' Version: ' + BrowserInfo.majorVersion;
    if (BrowserInfo.isFirefox()) {
        navigator?.gpu?.requestAdapter()
            .then( () => BrowserInfo.fireflyWebGpuEnabled= true)
            .catch(() => BrowserInfo.fireflyWebGpuEnabled= false);
    }
    BrowserInfo.platformMajorVerison= platformMajorVersion(ua,BrowserInfo.platform);
    logIfVersionNotSupported();
})();

function logIfVersionNotSupported() {
    if (isBrowserVersionSupported()) return;
    const minVer= getMinSupportedVersion();
    const bStr= BrowserInfo.getBrowserString();
    const platformStrStr= `(running on ${BrowserInfo.platform} ${BrowserInfo.platformMajorVerison})`;
    console.warn(`Browser not supported: ${BrowserInfo.browserDesc} is too old and not supported, the minimum ${bStr} version is ${minVer} ${platformStrStr}`);
}

function isBrowser(testBrowsers, version=UNKNOWN_VER ) {
    return testBrowsers===BrowserInfo.browser && (version===UNKNOWN_VER || version===BrowserInfo.majorVersion);
}

function isVersionAtLeast(version, minor= UNKNOWN_VER) {
    if (version===UNKNOWN_VER) return false;
    if (BrowserInfo.majorVersion>version) return true;
    if (BrowserInfo.majorVersion===version) {
        if (BrowserInfo.minorVersion===UNKNOWN_VER) return true;
        if (BrowserInfo.minorVersion!==UNKNOWN_VER && BrowserInfo.minorVersion>=minor) return true;
    }
    return false;
}

function isBrowserVersionSupported() {
    return BrowserInfo.isVersionAtLeast(BrowserInfo.getMinSupportedVersion());
}

function getMinSupportedVersion() {
    if (BrowserInfo.isChrome()) return Number(getProp('MIN_CHROME_VERSION','0'));
    if (BrowserInfo.isSafari()) return Number(getProp('MIN_SAFARI_VERSION','0'));
    if (BrowserInfo.isFirefox()) return Number(getProp('MIN_FIREFOX_VERSION','0'));
    if (BrowserInfo.isEdge()) return Number(getProp('MIN_EDGE_VERSION','0'));
    return 0;
}

function evaluateBrowser(ua,p) {
    const SEAMONKEY_KEY= 'seamonkey/';
    const FIREFOX_KEY= 'firefox/';
    const SIMPLE_VERSION_KEY = 'version/';
    const CHROME_KEY= 'chrome/';
    const OPERA_KEY= 'opr/';
    const EDGE_KEY= 'edg/';
    const IE_11_KEY= 'rv:';
    const unknown= {browser:Browser.UNKNOWN,majorVersion:UNKNOWN_VER, minorVersion:UNKNOWN_VER};
    if (!ua) return unknown;
    if (ua.includes('trident') && !ua.includes('edge')) {// IE is deprecated, but I am leaving the code here for now
        return parseVersion(ua,IE_11_KEY,Browser.IE);
    }
    else if (ua.includes(FIREFOX_KEY)) {
        return parseVersion(ua,FIREFOX_KEY,Browser.FIREFOX);
    }
    else if (ua.includes(SEAMONKEY_KEY)) { // as of fall 2025, SeaMonkey is still around
        return parseVersion(ua,SEAMONKEY_KEY,Browser.SEAMONKEY);
    }
    else if  (ua.includes('applewebkit')) {
        if (ua.includes(EDGE_KEY)) {
            return parseVersion(ua,EDGE_KEY,Browser.EDGE);
        }
        else if  (ua.includes(OPERA_KEY)) {
            return parseVersion(ua,OPERA_KEY,Browser.OPERA);
        }
        if (ua.includes(CHROME_KEY)) {
            return parseVersion(ua,CHROME_KEY,Browser.CHROME);
        }
        else if (p===Platform.BLACKBERRY || p===Platform.SYMBIAN_OS || p===Platform.UNKNOWN) {
            return parseVersion(ua,SIMPLE_VERSION_KEY,Browser.WEBKIT_GENERIC);
        }
        else {
            return parseVersion(ua,SIMPLE_VERSION_KEY,Browser.SAFARI);
        }
    }
    return unknown;
}

/**
 * @param ua user agent string
 * @param key
 * @param browser
 * @return {{majorVersion: number, minorVersion: number, browser}}
 */
function parseVersion(ua,key,browser) {
    let idx= ua.indexOf(key);
    if (idx!==-1) idx+= key.length;
    const ver= ua.substring(idx).match(/[0-9.]+/)?.[0];
    const sAry= ver?.split('\.') ?? [];
    return {
        browser,
        majorVersion: !isNaN(Number(sAry[0])) ? parseInt(sAry[0]) : UNKNOWN_VER,
        minorVersion: !isNaN(Number(sAry[1])) ? parseInt(sAry[1]) : UNKNOWN_VER
    };
}

function platformMajorVersion(ua,platform) {
    if (!ua) return 0;
    const osPart= ua.substring(ua.indexOf('(')+1, ua.indexOf(')'))?.split(';')[1];
    const osParts= osPart.split(/\s+/);
    const ver= osParts[osParts.length-1];
    const [majorVer,minorVer]= ver.split(/[._]/);
    return platform===Platform.MACOS ? Number(minorVer) : Number(majorVer);
}

function evaluatePlatform(ua) {
    if (!ua) return Platform.UNKNOWN;
    if  (ua.includes('window')) return Platform.WINDOWS;
    else if  (ua.includes('macintosh')) return Platform.MACOS;
    else if  (ua.includes('ipad')) return Platform.IPAD;
    else if  (ua.includes('iphone')) return Platform.IPHONE;
    else if  (ua.includes('linux')) return (ua.includes('android')) ? Platform.ANDROID : Platform.LINUX;
    else if  (ua.includes('solaris')) return Platform.SOLARIS;
    else if  (ua.includes('sunos')) return Platform.SUNOS;
    else if  (ua.includes('aix')) return Platform.AIX;
    else if  (ua.includes('hpux')) return Platform.HPUX;
    else if  (ua.includes('freebsd')) return Platform.FREE_BSD;
    else if  (ua.includes('symbianos')) return Platform.SYMBIAN_OS;
    else if  (ua.includes('j2me')) return Platform.J2ME;
    else if  (ua.includes('blackberry')) return Platform.BLACKBERRY;
    else return Platform.UNKNOWN;
}


function supportsCssColorMix() {
    if (isBrowser(Browser.CHROME) && isVersionAtLeast(111)) return true;
    if (isBrowser(Browser.SAFARI) && isVersionAtLeast(16,2)) return true;
    if (isBrowser(Browser.FIREFOX) && isVersionAtLeast(121)) return true;
    if (isBrowser(Browser.IE) && isVersionAtLeast(111)) return true;
    return false;
}

function supportsWebGpu() {
    const foundGpu= Boolean(navigator.gpu);
    if (isBrowser(Browser.CHROME)) return isVersionAtLeast(113) && foundGpu;
    if (isBrowser(Browser.EDGE)) return isVersionAtLeast(113) && foundGpu;
    if (isBrowser(Browser.SAFARI)) return isVersionAtLeast(26) && foundGpu;
    if (isBrowser(Browser.FIREFOX)) return isVersionAtLeast(141) && foundGpu && Boolean(BrowserInfo.fireflyWebGpuEnabled);
    if (isBrowser(Browser.OPERA)) return isVersionAtLeast(99) && foundGpu;
    return foundGpu;
}